yt_dlp/networking/utils.py

```diff
@@ -0,0 +1,192 @@
+    # Work around the issue in load_default_certs when there are bad certificates. See:
+    # https://github.com/yt-dlp/yt-dlp/issues/1060,
+    # https://bugs.python.org/issue35665, https://bugs.python.org/issue45312
+    except ssl.SSLError:
+        # enum_certificates is not present in mingw python. See https://github.com/yt-dlp/yt-dlp/issues/1151
+        if sys.platform == 'win32' and hasattr(ssl, 'enum_certificates'):
+            for storename in ('CA', 'ROOT'):
+                _ssl_load_windows_store_certs(context, storename)
+        context.set_default_verify_paths()
+
+
+def _ssl_load_windows_store_certs(ssl_context, storename):
+    # Code adapted from _load_windows_store_certs in https://github.com/python/cpython/blob/main/Lib/ssl.py
+    try:
+        certs = [cert for cert, encoding, trust in ssl.enum_certificates(storename)
+                 if encoding == 'x509_asn' and (
+                     trust is True or ssl.Purpose.SERVER_AUTH.oid in trust)]
+    except PermissionError:
+        return
+    for cert in certs:
+        with contextlib.suppress(ssl.SSLError):
+            ssl_context.load_verify_locations(cadata=cert)
+
+
+def socks_create_proxy_args(socks_proxy):
+    url_components = compat_urlparse.urlparse(socks_proxy)
+    if url_components.scheme.lower() == 'socks5':
+        socks_type = ProxyType.SOCKS5
+    elif url_components.scheme.lower() in ('socks', 'socks4'):
+        socks_type = ProxyType.SOCKS4
+    elif url_components.scheme.lower() == 'socks4a':
+        socks_type = ProxyType.SOCKS4A
+
+    def unquote_if_non_empty(s):
+        if not s:
+            return s
+        return compat_urllib_parse_unquote_plus(s)
+    return {
+        'proxytype': socks_type,
+        'addr': url_components.hostname,
+        'port': url_components.port or 1080,
+        'rdns': True,
+        'username': unquote_if_non_empty(url_components.username),
+        'password': unquote_if_non_empty(url_components.password),
+    }
+
+
+def select_proxy(url, proxies):
+    """Unified proxy selector for all backends"""
+    if proxies is None:
+        proxies = {}
+    url_components = urllib.parse.urlparse(url)
+    priority = [
+        url_components.scheme or 'http',  # prioritise more specific mappings
+        'all'
+    ]
+    return next((proxies[key] for key in priority if key in proxies), None)
+
+
+# Get a copy of std headers, while also retaining backwards compat with utils.std_headers
+# TODO: just make std_headers backwards compat with this
+def make_std_headers():
+    return CaseInsensitiveDict(_std_headers, std_headers)
+
+
+def update_request(req: Request, url: str = None, data=None,
+                   headers: typing.Mapping = None, query: dict = None):
+    """
+    Creates a copy of the request and updates relevant fields
+    """
+    req = req.copy()
+    req.data = data or req.data
+    req.headers.update(headers or {})
+    req.url = update_url_query(url or req.url, query or {})
+    return req
```

Comment:
You meaning something like this maybe?:
```py
def update(self, url=None, data=None, headers=None, query=None)
```

Reply:
https://github.com/yt-dlp/yt-dlp/pull/2861/commits/7a92aae218fcddbf35c7659afde918641260dd9c

Reply:
Actually having second thoughts about this implementation. Now instead of `update_request`, to get the same behaviour, we have to do:

```py
req = request.copy()
req.update(...)
```

Maybe could have some type of `new_child(...)` class method that does `update_request` exactly.

Reply:
update_request is only used in 2 places. And in both of those, I don't think returning a copy is important